client-misk-jooq/src/main/kotlin/app/cash/backfila/client/misk/jooq/JooqBackfill.kt

```diff
@@ -0,0 +1,147 @@
+
+  /**
+   * The table containing the rows the backfill will run over.
+   */
+  val table: TableLike<*>
+
+  /**
+   * Jooq condition that limits which rows the backfill runs over.
+   */
+  val filterCondition: Condition
+
+  /**
+   * List of fields that uniquely identifies keys to backfill.
+   *
+   * Note: this could be a list of a single field when you aren't using a real compound key.
+   */
+  val compoundKeyFields: List<Field<*>>
+
+  /**
+   * Convert from a JOOQ Record to the K type.
+   */
+  fun recordToKey(record: Record): K
+
+  /**
+   * Convert from the K type to a JOOQ Record.
+   */
+  fun keyToRecord(key: K): Record
+
+  /**
+   * A serializer for the configured key type, since the Backfila protos use [ByteString]s.
+   */
+  val keySerializer: ByteStringSerializer<K>
+
+  /**
+   * Hook that gives you a place to prepare or validate the backfill. Throw an exception from your
+   * consumer to indicate the backfill is invalid.
+   */
+  fun prepareAndValidateBackfill(config: BackfillConfig<Param>) {}
+
+  /**
+   * Function that performs the work of the backfill by applying some side effect to the given
+   * list of key values.
+   *
+   * Note: this will be called outside of a transaction.
+   */
+  fun backfill(backfillBatch: BackfillBatch<K, Param>)
+
+
+  fun toByteString(key: K): ByteString {
+    return keySerializer.toByteString(key)
+  }
+
+  fun fromByteString(byteString: ByteString?): K {
+    return keySerializer.fromByteString(byteString)
+  }
+
+  fun buildKeyRange(start: K, end: K): KeyRange {
+    return KeyRange.Builder()
+      .start(toByteString(start))
+      .end(toByteString(end))
+      .build()
+  }
+
+  fun <T> inTransactionReturning(
+    comment: String,
+    partitionName: String?,
+    work: (dslContext: DSLContext) -> T
+  ): T {
+    val transacterBackfill: BackfillJooqTransacter = getTransacter(partitionName)
+    return transacterBackfill.transaction(comment, work)
+  }
+
+  fun sortingByCompoundKeyFields(
+    withSortDirection: (field: Field<*>) -> SortField<*>
+  ): List<SortField<*>> {
+    return compoundKeyFields
+      .map(withSortDirection)
+  }
+
+  fun getUnfilteredBoundaryKeyValue(
+    partitionName: String,
+    withSortDirection: (field: Field<*>) -> SortField<*>
+  ): K? {
+    return inTransactionReturning(
+      "JooqConfig#getUnfilteredBoundaryKeyValue",
+      partitionName
+    ) { session: DSLContext ->
+      session
+        .select(compoundKeyFields)
+        .from(table)
+        .orderBy(sortingByCompoundKeyFields(withSortDirection))
+        .limit(1)
+        .fetchOptional {
+          recordToKey(it)
+        }.orElse(null)
+    }
+  }
+
+  fun getTransacter(partitionName: String?): BackfillJooqTransacter {
+    val transacterBackfill: BackfillJooqTransacter? = shardedTransacterMapBackfill[partitionName]
+    Preconditions.checkArgument(
+      transacterBackfill != null,
+      String.format(
```

Comment:
in kotlin, you can do `"not found: $partitionName"`

Reply:
I changed that to the below:
```
fun getTransacter(partitionName: String?): BackfillJooqTransacter {
    return shardedTransacterMapBackfill[partitionName]
      ?: throw IllegalStateException(
        "A JooqTransacter for the following partitionName was not found $partitionName"
      )
  }
  ```